src/rdkafka_broker.c

```diff
@@ -3051,20 +3059,49 @@ static void rd_kafka_broker_internal_serve (rd_kafka_broker_t *rkb,
 
 
 /**
- * @brief Scan toppar's xmit queue for message timeouts.
- * @locality broker thread
- * @locks none
+ * @brief Scan toppar's xmit and producer queue for message timeouts and
+ *        enqueue delivery reports for timed out messages.
+ *
+ * @returns the number of messages timed out.
+ *
+ * @locality toppar's broker handler thread
+ * @locks toppar_lock MUST be held
```

Comment:
I'm a bit confused by this annotation. It's unclear to me if this intended to inform the developer to acquire this lock prior to calling this function or this function acquires the lock. I ask because the former appears to be the case for this function however `rd_kafka_toppar_producer_serve` has the same annotation yet it locks the toppar shortly after entering the function. Sure you wouldn't want to acquire the lock prior to calling `rd_kafka_toppar_producer_serve` otherwise you'd end up in a deadlock. 

annotation
https://github.com/edenhill/librdkafka/pull/2188/files#diff-214080cd229adac5b1596529b6eecfe4R3129

grab lock:
https://github.com/edenhill/librdkafka/pull/2188/files#diff-214080cd229adac5b1596529b6eecfe4R3154

There is a subtle difference in the annotations, perhaps that's intentional.